pkg/util/snmp.go

```diff
@@ -0,0 +1,144 @@
+	ADIdentifier       string   `mapstructure:"ad_identifier"`
+}
+
+// Digest returns an hash value representing the data stored in this configuration, minus the network address
+func (c *SNMPConfig) Digest(address string) string {
+	h := fnv.New64()
+	h.Write([]byte(address))
+	h.Write([]byte(fmt.Sprintf("%d", c.Port)))
+	h.Write([]byte(c.Version))
+	h.Write([]byte(c.Community))
+	h.Write([]byte(c.User))
+	h.Write([]byte(c.AuthKey))
+	h.Write([]byte(c.AuthProtocol))
+	h.Write([]byte(c.PrivKey))
+	h.Write([]byte(c.PrivProtocol))
+	h.Write([]byte(c.ContextEngineID))
+	h.Write([]byte(c.ContextName))
+	for _, ip := range c.IgnoredIPAddresses {
+		h.Write([]byte(ip))
+	}
+
+	return strconv.FormatUint(h.Sum64(), 16)
+}
+
+// BuildSNMPParams returns a valid GoSNMP struct to start making queries
+func (c *SNMPConfig) BuildSNMPParams() (*gosnmp.GoSNMP, error) {
+	port := c.Port
+	if port == 0 {
+		port = 161
+	}
+
+	if c.Community == "" && c.User == "" {
+		return nil, errors.New("No authentication mechanism specified")
+	}
+
+	var version gosnmp.SnmpVersion
+	if c.Version == "1" {
+		version = gosnmp.Version1
+	} else if c.Version == "2" || (c.Version == "" && c.Community != "") {
+		version = gosnmp.Version2c
+	} else if c.Version == "3" || (c.Version == "" && c.User != "") {
+		version = gosnmp.Version3
+	} else {
+		return nil, fmt.Errorf("SNMP version not supported: %s", c.Version)
+	}
+
+	var authProtocol gosnmp.SnmpV3AuthProtocol
+	if c.AuthProtocol == "MD5" {
+		authProtocol = gosnmp.MD5
+	} else if c.AuthProtocol == "SHA" {
+		authProtocol = gosnmp.SHA
+	} else if c.AuthProtocol != "" {
+		return nil, fmt.Errorf("Unsupported authentication protocol: %s", c.AuthProtocol)
+	}
+
+	var privProtocol gosnmp.SnmpV3PrivProtocol
+	if c.PrivProtocol == "DES" {
+		privProtocol = gosnmp.DES
+	} else if c.PrivProtocol == "AES" {
+		privProtocol = gosnmp.AES
+	} else if c.PrivProtocol != "" {
+		return nil, fmt.Errorf("Unsupported privacy protocol: %s", c.PrivProtocol)
+	}
+
+	msgFlags := gosnmp.NoAuthNoPriv
+	if c.PrivKey != "" {
+		msgFlags = gosnmp.AuthPriv
+	} else if c.AuthKey != "" {
+		msgFlags = gosnmp.AuthNoPriv
+	}
+
+	return &gosnmp.GoSNMP{
+		Port:            port,
+		Community:       c.Community,
+		Transport:       "udp",
+		Version:         version,
+		Timeout:         time.Duration(c.Timeout) * time.Second,
+		Retries:         c.Retries,
+		SecurityModel:   gosnmp.UserSecurityModel,
+		MsgFlags:        msgFlags,
+		ContextEngineID: c.ContextEngineID,
+		ContextName:     c.ContextName,
+		SecurityParameters: &gosnmp.UsmSecurityParameters{
+			UserName:                 c.User,
+			AuthenticationProtocol:   authProtocol,
+			AuthenticationPassphrase: c.AuthKey,
+			PrivacyProtocol:          privProtocol,
+			PrivacyPassphrase:        c.PrivKey,
+		},
+	}, nil
+}
+
+// IsIPIgnored checks the given IP against IgnoredIPAddresses
+func (c *SNMPConfig) IsIPIgnored(ip net.IP) bool {
+	ipString := ip.String()
+	for _, ip := range c.IgnoredIPAddresses {
```

Comment:
For perf, wdyta making it a map after parsing? (Go has no sets)

Reply:
Have you look at the [`DecodeHookFunc`](https://github.com/mitchellh/mapstructure/blob/14428cd8e72132155ec2f0f13c4eb93ffe620f7c/mapstructure.go#L160) provided by mapstructure package.
it looks to be possible to convert `[]string` type to `map[string]interface{}` on the fly when the configuration is decoded.
An example of the `DecodeHoookFunc` usage: https://github.com/mitchellh/mapstructure/blob/ce2ff0c13ce509e36e9254c08ea0bca90ed5af6c/decode_hooks_test.go#L86

Reply:
Yep tried something.